File: app/src/main/java/com/afoxxvi/alopex/service/AlopexNotificationListenerService.java
package com.afoxxvi.alopex.service;

import static com.afoxxvi.alopex.AlopexView.TAG;
import static com.afoxxvi.alopex.MainActivity.CHANNEL_ONE_ID;

import android.app.Notification;
import android.content.Intent;
import android.graphics.drawable.Icon;
import android.os.Bundle;
import android.os.Message;
import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;
import android.util.Log;

import androidx.core.app.NotificationManagerCompat;

import com.afoxxvi.alopex.AlopexView;
import com.afoxxvi.alopex.R;
import com.afoxxvi.alopex.component.filter.AlopexFilterManager;
import com.afoxxvi.alopex.ui.fragment.NotificationFragment;
import com.afoxxvi.alopex.util.Pair;
import com.afoxxvi.alopex.util.Triplet;

import java.util.HashMap;
import java.util.Map;

public class AlopexNotificationListenerService extends NotificationListenerService {
    public static boolean active = false;
    private static final Map<String, Pair<String, String>> LATEST_NOTIFICATION_MAP = new HashMap<>();
    public static final String PROPERTY_KEY_LAST_NOTIFICATION_MILLS = "lastNotificationMills";
    private static long lastNotificationMills = 0;

    public AlopexNotificationListenerService() {
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    private void handleNotification(StatusBarNotification sbn) {
        Bundle bundle = sbn.getNotification().extras;
        String pkg = sbn.getPackageName();
        if (pkg.equals(getApplication().getPackageName())) {
            return;
        }
        String title = bundle.getString(Notification.EXTRA_TITLE, "");
        String text = bundle.getString(Notification.EXTRA_TEXT, "");
        if (LATEST_NOTIFICATION_MAP.containsKey(pkg)) {
            Pair<String, String> old = LATEST_NOTIFICATION_MAP.get(pkg);
            if (old != null && old.a.equals(title) && old.b.equals(text)) {
                return;
            }
        }
        LATEST_NOTIFICATION_MAP.put(pkg, new Pair<>(title, text));
        if (AlopexView.handlerNotification != null) {
            Message msg = Message.obtain(AlopexView.handlerNotification, NotificationFragment.WHAT_NOTIFICATION_LISTENER_SERVICE);
            Bundle bd = new Bundle();
            bd.putString("package", pkg);
            bd.putString("title", title);
            bd.putString("content", text);
            bd.putLong("time", sbn.getPostTime());
            msg.setData(bd);
            msg.sendToTarget();
        }
        Triplet<Boolean, Boolean, Boolean> pair = AlopexFilterManager.getInstance().isFiltered(pkg, title, text, true);
        if (pair.a) {
            Notification ntf = sbn.getNotification();
            Icon icon = ntf.getSmallIcon();
            Icon large = ntf.getLargeIcon();
            Notification notification = new Notification.Builder(this, CHANNEL_ONE_ID)
                    .setAutoCancel(true)
                    .setContentTitle(title)
                    .setContentText(text)
                    .setContentIntent(ntf.contentIntent)
                    .setSmallIcon(icon)
                    .setLargeIcon(large)
                    //.setTimeoutAfter(1000)
                    .build();
            NotificationManagerCompat managerCompat = NotificationManagerCompat.from(this);
            managerCompat.notify(1, notification);
        }
        if (pair.b) {
            cancelNotification(sbn);
        }
    }

    private void saveProperties() {
        AlopexView.getProperties().edit().putLong(PROPERTY_KEY_LAST_NOTIFICATION_MILLS, lastNotificationMills).apply();
    }

    @Override
    public void onNotificationPosted(StatusBarNotification sbn) {
        lastNotificationMills = sbn.getPostTime();
        saveProperties();
        handleNotification(sbn);
    }

    @Override
    public void onListenerConnected() {
        super.onListenerConnected();
        lastNotificationMills = AlopexView.getProperties().getLong(PROPERTY_KEY_LAST_NOTIFICATION_MILLS, 0);
        Log.i(TAG, "onListenerConnected: ");
        active = true;
        Notification notification = new Notification.Builder(this, CHANNEL_ONE_ID)
                .setAutoCancel(true)
                .setContentTitle("Listener Connected")
                .setContentText("at " + AlopexView.getDateTimeString())
                .setSmallIcon(R.drawable.ic_launcher_foreground)
                .build();
        NotificationManagerCompat managerCompat = NotificationManagerCompat.from(this);
        managerCompat.notify(1000, notification);
        //startForeground(1, notification);
        AlopexView.notifyConnect();
        long maxMills = lastNotificationMills;
        for (StatusBarNotification activeNotification : getActiveNotifications()) {
            if (activeNotification.getPostTime() > lastNotificationMills) {
                handleNotification(activeNotification);
                maxMills = Math.max(maxMills, activeNotification.getPostTime());
            }
        }
        lastNotificationMills = maxMills;
        saveProperties();
    }

    @Override
    public void onListenerDisconnected() {
        super.onListenerDisconnected();
        Log.i(TAG, "onListenerDisconnected: ");
        active = false;
        Notification notification = new Notification.Builder(this, CHANNEL_ONE_ID)
                .setAutoCancel(true)
                .setContentTitle("Listener Disconnected")
                .setContentText("at " + AlopexView.getDateTimeString())
                .setSmallIcon(R.drawable.ic_launcher_foreground)
                .build();
        NotificationManagerCompat managerCompat = NotificationManagerCompat.from(this);
        managerCompat.notify(1000, notification);
        AlopexView.notifyDisconnect();
    }

    public void cancelNotification(StatusBarNotification sbn) {
        cancelNotification(sbn.getKey());
    }
}

File: app/src/main/java/com/afoxxvi/alopex/service/AlopexNotificationListenerService.kt
package com.afoxxvi.alopex.service

import android.Manifest
import android.annotation.SuppressLint
import android.app.Notification
import android.app.PendingIntent
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Message
import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import android.util.Log
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationManagerCompat
import com.afoxxvi.alopex.Alopex
import com.afoxxvi.alopex.MainActivity
import com.afoxxvi.alopex.R
import com.afoxxvi.alopex.component.filter.Filters
import com.afoxxvi.alopex.component.notify.Notifications
import com.afoxxvi.alopex.component.notify.WrapperNotification
import com.afoxxvi.alopex.ui.fragment.NotificationFragment
import com.afoxxvi.alopex.util.FoxTools

class AlopexNotificationListenerService : NotificationListenerService() {
    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        return START_STICKY
    }

    private fun sendNotification(notification: Notification, id: Int) {
        val managerCompat = NotificationManagerCompat.from(this)
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            return
        }
        managerCompat.notify(id, notification)
    }

    private fun handleNotification(sbn: StatusBarNotification) {
        val bundle = sbn.notification.extras
        val pkg = sbn.packageName
        if (pkg == application.packageName) return
        val title = bundle.getString(Notification.EXTRA_TITLE, "")
        val text = bundle.getString(Notification.EXTRA_TEXT, "")
        if (LATEST_NOTIFICATION_MAP.containsKey(pkg)) {
            val old = LATEST_NOTIFICATION_MAP[pkg]
            if (old != null && old.first == title && old.second == text) {
                return
            }
        }
        LATEST_NOTIFICATION_MAP[pkg] = Pair(title, text)
        Alopex.handlerNotification?.let {
            val msg = Message.obtain(it, NotificationFragment.WHAT_NOTIFICATION_LISTENER_SERVICE)
            val wrapperNotification = WrapperNotification(title, text, FoxTools.getLocalDateTimeFromMills(System.currentTimeMillis()))
            val fromIndex = Notifications.newNotify(applicationContext, pkg, wrapperNotification)
            msg.arg1 = fromIndex
            msg.sendToTarget()
        }
        val result = Filters.passNotification(pkg, title, text)
        if (result.doNotify) {
            val ntf = sbn.notification
            val icon = ntf.smallIcon
            val large = ntf.getLargeIcon()
            val notification = Notification.Builder(this, Alopex.CHANNEL_ID)
                .setAutoCancel(true)
                .setContentTitle(title)
                .setContentText(text)
                .setContentIntent(ntf.contentIntent)
                .setSmallIcon(icon)
                .setLargeIcon(large) //.setTimeoutAfter(1000)
                .build()
            sendNotification(notification, 1000)
        }
        if (result.doCancel) {
            cancelNotification(sbn.key)
        }
    }

    private fun saveProperties() {
        Alopex.properties!!.edit().putLong(PROPERTY_KEY_LAST_NOTIFICATION_MILLS, lastNotificationMills).apply()
    }

    override fun onNotificationPosted(sbn: StatusBarNotification) {
        lastNotificationMills = sbn.postTime
        saveProperties()
        handleNotification(sbn)
    }

    @SuppressLint("UnspecifiedImmutableFlag")
    override fun onListenerConnected() {
        super.onListenerConnected()
        lastNotificationMills = Alopex.properties!!.getLong(PROPERTY_KEY_LAST_NOTIFICATION_MILLS, 0)
        Log.i(Alopex.TAG, "onListenerConnected: ")
        active = true
        val pendingIntent: PendingIntent = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            PendingIntent.getActivity(
                this, 0, Intent(this, MainActivity::class.java), PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_MUTABLE
            )
        } else {
            PendingIntent.getActivity(this, 0, Intent(this, MainActivity::class.java), PendingIntent.FLAG_UPDATE_CURRENT)
        }
        val notification =
            Notification.Builder(this, Alopex.CHANNEL_ID)
                .setAutoCancel(false)
                .setOngoing(true)
                .setContentTitle("Alopex is running")
                .setContentIntent(pendingIntent)
                .setSmallIcon(R.drawable.ic_launcher_foreground)
                .build()
        sendNotification(notification, 1)
        startForeground(1, notification)
        Alopex.notifyConnect()
        var maxMills = lastNotificationMills
        for (activeNotification in activeNotifications) {
            if (activeNotification.postTime > lastNotificationMills) {
                handleNotification(activeNotification)
                maxMills = maxMills.coerceAtLeast(activeNotification.postTime)
            }
        }
        lastNotificationMills = maxMills
        saveProperties()
    }

    override fun onListenerDisconnected() {
        super.onListenerDisconnected()
        Alopex.notifyDisconnect()
    }

    companion object {
        var active = false
        private val LATEST_NOTIFICATION_MAP: MutableMap<String, Pair<String, String>> = HashMap()
        const val PROPERTY_KEY_LAST_NOTIFICATION_MILLS = "lastNotificationMills"
        private var lastNotificationMills: Long = 0
    }
}